[DataPipes] Add group support to the sharding_filter

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88424

Differential Revision: [D41006747](https://our.internmc.facebook.com/intern/diff/D41006747)

VitalyFedyunin committed:
[DataPipes] Add group support to the sharding_filter

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataPipes] Add group support to the sharding_filter"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataPipes] Add group support to the sharding_filter"




[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataPipes] Add group support to the sharding_filter"




[ghstack-poisoned]

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -25,17 +32,45 @@ class ShardingFilterIterDataPipe(IterDataPipe):
+
+    def __init__(self, source_datapipe: IterDataPipe, sharding_group_filter=None):
         self.source_datapipe = source_datapipe
+        self.sharding_group_filter = sharding_group_filter
```
Do we need an extra API to set `sharding_group_filter`?

Based on the implementation, it seems `sharding_group_filter` is an integer, could we change it `set` or `list` to support multiple filters?

VitalyFedyunin replied:
I have no use-cases for it, but it will be trivial to change later if we require to.

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -25,17 +32,45 @@ class ShardingFilterIterDataPipe(IterDataPipe):
+    def _update_num_of_instances(self):
+        sorted_sharding_groups = []
+        for key in sorted(self.groups.keys()):
+            if self.sharding_group_filter is None:
+                sorted_sharding_groups.append(self.groups[key])
+            else:
+                if key == self.sharding_group_filter:
+                    sorted_sharding_groups.append(self.groups[key])
```
nit:
```py
if self.sharding_group_filter is None or key == self.sharding_group_filter:
    sorted_sharding_groups.append(self.groups[key])
```

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -25,17 +32,45 @@ class ShardingFilterIterDataPipe(IterDataPipe):
-    def apply_sharding(self, num_of_instances, instance_id):
-        self.num_of_instances = num_of_instances
-        self.instance_id = instance_id
+    def apply_sharding(self, num_of_instances, instance_id, sharding_group=SHARDING_PRIORITIES.DEFAULT):
```
Super nit: Could we add a validation that `instance_id < num_of_instances`?

ejguan commented on test/test_datapipe.py:
```diff
@@ -2663,6 +2667,30 @@ def test_simple_sharding(self):
+
+        with self.assertRaises(Exception):
+            dp.apply_sharding(2, 1, sharding_group=SHARDING_PRIORITIES.DEFAULT)
+            dp.apply_sharding(5, 3, sharding_group=SHARDING_PRIORITIES.MULTIPROCESSING)
```
Could we add a separate context of `self.assertRaises` for the second Error?

ejguan reviewed (approved):
LGTM with a few nit comments.

VitalyFedyunin committed:
Update on "[DataPipes] Add group support to the sharding_filter"




[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41006747).

VitalyFedyunin committed:
Update on "[DataPipes] Add group support to the sharding_filter"


Differential Revision: [D41006747](https://our.internmc.facebook.com/intern/diff/D41006747)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "[DataPipes] Add group support to the sharding_filter"


Differential Revision: [D41006747](https://our.internmc.facebook.com/intern/diff/D41006747)

[ghstack-poisoned]

VitalyFedyunin commented:
@VitalyFedyunin has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D41006747).

VitalyFedyunin commented:
/easycla

VitalyFedyunin commented:
/easycla

wenleix commented on torch/utils/data/graph_settings.py:
```diff
@@ -42,7 +46,7 @@ def apply_sharding(datapipe: DataPipe, num_of_instances: int, instance_id: int)
                         raise RuntimeError('This implementation of sharding can be only applied once per instance of DataPipeline.',
                                            'Already applied to', already_applied_to, 'while trying to apply to', pipe)
-                    pipe.apply_sharding(num_of_instances, instance_id)
+                    pipe.apply_sharding(num_of_instances, instance_id, sharding_group=sharding_group)
```
noob question: does `is_shardable` and `apply_sharding` only exists in `ShardingFilterIterDataPipe` ? 

Also if there is no `ShardingFilterIterDataPipe`, look like no sharding will happen? (shall we error in that case? ) :) 

cc @ejguan
